File: backend/schemas/business.py
"""Pydantic schemas for the Dream-to-Launch Builder endpoints."""
from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ── Quiz ──────────────────────────────────────────────────────────────────────

SKILLS_OPTIONS = [
    "computers_tech", "sales_marketing", "writing_content", "design_creative",
    "teaching_coaching", "cooking_food", "trades_repair", "healthcare_wellness",
    "finance_accounting", "management_leadership", "customer_service", "languages",
    "music_arts", "sports_fitness", "childcare_education",
]

PROBLEMS_OPTIONS = [
    "save_time", "save_money", "reduce_stress", "learn_something",
    "improve_health", "find_community", "get_entertainment", "solve_tech_problem",
    "improve_home", "grow_business", "get_professional_services", "get_local_services",
]

BUSINESS_TYPE_OPTIONS = Literal["service", "product", "online", "local"]
CAPITAL_OPTIONS = Literal["<500", "500-2k", "2k-10k", "10k+"]
HOURS_OPTIONS = Literal["<5", "5-15", "15-30", "30+"]


class QuizSubmitRequest(BaseModel):
    skills: list[str] = Field(..., min_length=1, max_length=15)
    problems: list[str] = Field(..., min_length=1, max_length=12)
    business_type: BUSINESS_TYPE_OPTIONS
    starting_capital: CAPITAL_OPTIONS
    weekly_hours: HOURS_OPTIONS


class IdeaSuggestion(BaseModel):
    name: str
    description: str
    business_fit_pct: int = Field(..., ge=0, le=100)
    startup_cost_tier: Literal["Low", "Medium", "High"]
    difficulty_tier: Literal["Easy", "Medium", "Hard"]
    revenue_potential: str


class QuizSubmitResponse(BaseModel):
    business_idea_id: UUID
    suggestions: list[IdeaSuggestion]
    mission_preview: str


# ── Ideas ─────────────────────────────────────────────────────────────────────

class BusinessIdeaListItem(BaseModel):
    id: UUID
    selected_idea_index: Optional[int] = None
    business_fit_pct: Optional[int] = None
    startup_cost_tier: Optional[str] = None
    difficulty_tier: Optional[str] = None
    revenue_potential: Optional[str] = None
    status: str
    created_at: datetime
    idea_name: Optional[str] = None
    idea_description: Optional[str] = None


class BusinessIdeaListResponse(BaseModel):
    data: list[BusinessIdeaListItem]


class SaveIdeaRequest(BaseModel):
    idea_index: int = Field(..., ge=0, le=2)


class SaveIdeaResponse(BaseModel):
    business_idea_id: UUID
    idea_name: str
    idea_description: str
    business_fit_pct: int
    startup_cost_tier: str
    difficulty_tier: str
    revenue_potential: str
    mission_preview: str


# ── Plans ─────────────────────────────────────────────────────────────────────

class LaunchPlanResponse(BaseModel):
    business_idea_id: UUID
    tier: str
    checklist: Optional[list[dict[str, Any]]] = None
    cost_calculator: Optional[list[dict[str, Any]]] = None
    pricing_packages: Optional[list[dict[str, Any]]] = None
    thirty_day_plan: Optional[list[dict[str, Any]]] = None
    business_plan_text: Optional[str] = None
    mission_vision: Optional[str] = None
    customer_persona: Optional[dict[str, Any]] = None
    funding_checklist: Optional[list[dict[str, Any]]] = None
    cyber_ai_checklist: Optional[list[dict[str, Any]]] = None
    ninety_day_roadmap: Optional[list[dict[str, Any]]] = None
    pdf_url: Optional[str] = None
    created_at: datetime


class GeneratePdfResponse(BaseModel):
    pdf_url: str
    generated_at: datetime


# ── Advisor Request ───────────────────────────────────────────────────────────

class AdvisorRequestBody(BaseModel):
    business_idea_id: Optional[UUID] = None


class AdvisorRequestResponse(BaseModel):
    advisor_request_id: UUID
    status: str
    message: str

File: mobile/src/api/business.ts
import { api } from './client';

export interface IdeaSuggestion {
  name: string;
  description: string;
  business_fit_pct: number;
  startup_cost_tier: 'Low' | 'Medium' | 'High';
  difficulty_tier: 'Easy' | 'Medium' | 'Hard';
  revenue_potential: string;
}

export interface QuizSubmitResponse {
  business_idea_id: string;
  suggestions: IdeaSuggestion[];
  mission_preview: string;
}

export interface SavedIdeaData {
  business_idea_id: string;
  idea_name: string;
  idea_description: string;
  business_fit_pct: number;
  startup_cost_tier: string;
  difficulty_tier: string;
  revenue_potential: string;
  mission_preview: string;
}

export interface LaunchPlan {
  business_idea_id: string;
  tier: string;
  checklist: Array<{ item: string; category: string; required: boolean }> | null;
  cost_calculator: Array<{ item: string; estimated_cost_low: number; estimated_cost_high: number; category: string }> | null;
  pricing_packages: Array<{ name: string; description: string; price_suggestion: string; included_services: string[] }> | null;
  thirty_day_plan: Array<{ week: number; title: string; milestones: string[] }> | null;
  business_plan_text: string | null;
  mission_vision: string | null;
  customer_persona: Record<string, string> | null;
  funding_checklist: Array<{ item: string; category: string; completed: boolean }> | null;
  cyber_ai_checklist: Array<{ item: string; category: string; completed: boolean }> | null;
  ninety_day_roadmap: Array<{ month: number; title: string; goals: string[] }> | null;
  pdf_url: string | null;
  created_at: string;
}

export interface PurchaseStatusResponse {
  launch_builder: boolean;
  launch_packet_pro: boolean;
  advisor_review: boolean;
}

export interface OneTimeCheckoutResponse {
  client_secret: string;
  payment_intent_id: string;
  amount_cents: number;
  product_key: string;
}

export async function submitQuiz(
  token: string,
  payload: {
    skills: string[];
    problems: string[];
    business_type: string;
    starting_capital: string;
    weekly_hours: string;
  }
): Promise<QuizSubmitResponse> {
  return api.post<QuizSubmitResponse>('/business/quiz', payload, token);
}

export async function saveIdea(
  token: string,
  ideaId: string,
  ideaIndex: number
): Promise<SavedIdeaData> {
  return api.post<SavedIdeaData>(`/business/ideas/${ideaId}/save`, { idea_index: ideaIndex }, token);
}

export async function getLaunchPlan(
  token: string,
  ideaId: string,
  tier: string
): Promise<LaunchPlan> {
  return api.get<LaunchPlan>(`/business/ideas/${ideaId}/plan?tier=${tier}`, token);
}

export async function generatePdf(
  token: string,
  ideaId: string
): Promise<{ pdf_url: string; generated_at: string }> {
  return api.post<{ pdf_url: string; generated_at: string }>(
    `/business/ideas/${ideaId}/generate-pdf`,
    {},
    token
  );
}

export async function getPurchaseStatus(token: string): Promise<PurchaseStatusResponse> {
  return api.get<PurchaseStatusResponse>('/billing/purchases', token);
}

export async function createOneTimeCheckout(
  token: string,
  productKey: string
): Promise<OneTimeCheckoutResponse> {
  return api.post<OneTimeCheckoutResponse>(
    '/billing/one-time-checkout',
    { product_key: productKey },
    token
  );
}

export async function submitAdvisorRequest(
  token: string,
  businessIdeaId: string | null
): Promise<{ advisor_request_id: string; status: string; message: string }> {
  return api.post<{ advisor_request_id: string; status: string; message: string }>(
    '/business/advisor-request',
    { business_idea_id: businessIdeaId },
    token
  );
}
